File: packages/snow_draw_engine/lib/draw/utils/transforms/resize_anchor_point.dart
import '../../types/draw_point.dart';
import '../../types/draw_rect.dart';
import '../../types/resize_mode.dart';

/// Returns the opposite anchor point for a given resize handle [mode].
///
/// The returned point is in the same coordinate space as [rect] (typically the
/// selection overlay's un-rotated local frame).
DrawPoint oppositeBoundPointLocal(DrawRect rect, ResizeMode mode) =>
    switch (mode) {
      ResizeMode.topLeft => DrawPoint(x: rect.maxX, y: rect.maxY),
      ResizeMode.topRight => DrawPoint(x: rect.minX, y: rect.maxY),
      ResizeMode.bottomRight => DrawPoint(x: rect.minX, y: rect.minY),
      ResizeMode.bottomLeft => DrawPoint(x: rect.maxX, y: rect.minY),
      ResizeMode.top => DrawPoint(x: rect.centerX, y: rect.maxY),
      ResizeMode.bottom => DrawPoint(x: rect.centerX, y: rect.minY),
      ResizeMode.left => DrawPoint(x: rect.maxX, y: rect.centerY),
      ResizeMode.right => DrawPoint(x: rect.minX, y: rect.centerY),
    };

File: packages/snow_draw_rust_engine/src/draw/utils/transforms/resize_anchor_point.rs
#![allow(dead_code)]

use crate::draw::types::draw_point::DrawPoint;
use crate::draw::types::draw_rect::DrawRect;
use crate::draw::types::resize_mode::ResizeMode;

/// Returns the opposite anchor point for a resize handle in local bounds space.
///
/// The returned point is expressed in the same coordinate space as `rect`
/// (typically the selection overlay's un-rotated local frame).
pub fn opposite_bound_point_local(rect: DrawRect, mode: ResizeMode) -> DrawPoint {
    match mode {
        ResizeMode::TopLeft => DrawPoint::new(rect.max_x, rect.max_y),
        ResizeMode::TopRight => DrawPoint::new(rect.min_x, rect.max_y),
        ResizeMode::BottomRight => DrawPoint::new(rect.min_x, rect.min_y),
        ResizeMode::BottomLeft => DrawPoint::new(rect.max_x, rect.min_y),
        ResizeMode::Top => DrawPoint::new(rect.center_x(), rect.max_y),
        ResizeMode::Bottom => DrawPoint::new(rect.center_x(), rect.min_y),
        ResizeMode::Left => DrawPoint::new(rect.max_x, rect.center_y()),
        ResizeMode::Right => DrawPoint::new(rect.min_x, rect.center_y()),
    }
}
